### DISCUSSION TITLE
Development Security Considerations

### DISCUSSION THREAD

**Discussion by VisualMelon** [oxyplot/oxyplot#2029] [General]
The library is by its nature and design very 'low-risk': this isn't about issues with the library as-is; rather, certain recent events have me thinking about the development process, particularly if we are considering automatically pushing pre-release packages to nuget again (https://github.com/oxyplot/oxyplot/pull/1912#issuecomment-1659810397), and we should probably have a proper discussion about it. Obviously this is mostly a consideration for existing admins/maintainers, but other input welcome.

Without thinking too hard I'd suggest we change the configuration for both develop and master to
 - disallow pushing commits directly to the branches
 - require any PR must be approved by at least 1 maintainer who isn't the author

This would limit the harm that can be done by e.g. leaked credentials, and provide confidence that at least two people have looked at any piece of 'published' code. I think adding these 3 check-marks in the branch settings for each branch would achieve this:

![image](https://github.com/oxyplot/oxyplot/assets/5646475/49343d5c-1813-4567-970a-8ca8355fd94e)

The obvious concern with requiring this is that it means that we need two active maintainers to stop things grinding to a halt. I'm not intending to go anywhere, but as will be obvious to everyone I'm not as active as I'd like, though I can usually find time for reviews (which we have tacitly demanded anyway).

I would prefer that any PR requires 2 maintainers to review it (including the author), but that doesn't seem to be an option. (This is the only way to defeat a malicious maintainer, or now that I think about it leaked credentials for a maintainer beyond just the ability to push code)

**Comment by VisualMelon** [comment]
Mindless other thoughts whereabout I don't really know enough thereon to comment (list may grow over time):
 - it would be nice if a user can have some confidence that the packages on nuget come from this repo. I've pushed the last few releases manually which (apart from being error prone (see 2.1.1)) means we have to trust me and my computer.
 - Snyk seems to have stopped piping up, so need to check if we have an automatic mechanism for flagging dependency changes

**Comment by Jonarw** [comment]
Yes, thanks for raising this! I agree with all the points you made so far.
Additionally, I would suggest that we could consider requiring 2-factor authentication from all maintainers. This would at least somewhat reduce the probability of malicious code changes via a stolen maintainer account.

**Reply by VisualMelon** [reply]
I think 2FA is already a requirement for anyone pushing code (per https://github.blog/2023-03-09-raising-the-bar-for-software-security-github-2fa-begins-march-13/ ) but I could be misunderstanding that. Certainly agree it's a good idea if we can find a check-box for that.

**Reply by Jonarw** [reply]
Ah, wasn't aware of that.
There is this checkbox:
![image](https://github.com/oxyplot/oxyplot/assets/18097859/6ed5db91-b0cf-4c0d-8845-95e5f6a75752)

Though I guess it doesn't really make much of a difference, as pretty much all maintainers have 2FA enabled anyway.

Another aspect of this is of course: more people with write access of course also means more risk for security-related incidents (be it by accident or malicious third parties). Currently we have >10 people with write access to the core repo, and some of them have not been active in quite some time (as far as I know). I really don't want to open a can of worms here, and in no way do I want to imply that I distrust any of them. I also don't have any concrete suggestion as to if or how we would change that, just wanted to bring up this point as well.

**Reply by VisualMelon** [reply]
@objorke has mentioned this to me in the past, and previously I thought it wasn't a big deal having these older accounts 'on the books', but it's a much bigger deal if we have automatic push to nuget, so I would support pruning the maintainer list a bit: we can always add them back if they become active again.

**Reply by VisualMelon** [reply]
Can maybe add a new team for these older accounts with triage permissions.
![image](https://github.com/oxyplot/oxyplot/assets/5646475/71f5dcd6-8664-4161-af38-d536004d4f72)

**Reply by VisualMelon** [reply]
Finally found that checkbox: are we happy we the consideration that anyone who doesn't have 2FA will be removed?

**Reply by Jonarw** [reply]
I would support requiring 2FA from all maintainers. There is only one account who would be affected by this, and their last contribution to this repo seems to have been in 2015.

I also think creating a new team with triage permissions for maintainers who haven't been active in a long time might be a good idea - as you say, we can always add them back.

**Reply by VisualMelon** [reply]
@Jonarw when I 'press the button' it says there are 3 members to be removed. I'm happy for them to be removed, but I'll let you press the button on them since it's clearly not quite what you expected (not sure if all that information is public, so won't paste a screenshot here)

**Reply by Jonarw** [reply]
Ah, in addition to the one maintainer without 2FA there are 2 outside collaborateurs which I hadn't seen before. I'll 'press the button'.

**Comment by VisualMelon** [comment]
I've applied the "Review at least one reviewer" change to develop and master (obviously can change it back if there are objections). It looks like there is no way to achieve "Require at least 1 review if the author is maintainer, and require at least 2 reviews if the author is not".
